src/Psalm/Internal/Analyzer/StatementsAnalyzer.php

```diff
@@ -800,12 +816,201 @@ function ($line) {
+                    return $rhs_removable;
+                }
+            }
+            return $curr_removable;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * @param  array<PhpParser\Node\Stmt>   $stmts
+     * @param  string   $var_id
+     * @param  CodeLocation   $original_location
+     * @return array{0: PhpParser\Node\Stmt|null, 1:
+         PhpParser\Node\Expr\Assign|PhpParser\Node\Expr\AssignOp|PhpParser\Node\Expr\AssignRef|null}
+     */
+    private function findAssignStmt(array $stmts, string $var_id, CodeLocation $original_location)
+    {
+        $assign_stmt = null;
+        $assign_exp = null;
+        $assign_exp_found = false;
+        
+        $i = 0;
+        while ($i<count($stmts) && !$assign_exp_found) {
+            $stmt = $stmts[$i];
+            if ($stmt instanceof PhpParser\Node\Stmt\Expression) {
+                $search_result = $this->findAssignExp($stmt->expr, $var_id, $original_location->raw_file_start);
+                $target_exp = $search_result[0];
+                $levels_taken = $search_result[1];
+
+                if (!is_null($target_exp)) {
+                    $assign_exp_found = true;
+                    $assign_exp = $target_exp;
+                    $assign_stmt = $levels_taken === 1 ? $stmt : null;
+                }
+            }
+            $i+=1;
+        }
+
+        return [$assign_stmt, $assign_exp];
+    }
+
+    /**
+     * @param  PhpParser\Node\Expr $current_node
+     * @param  string   $var_id
+     * @param  int      $var_start_loc
+     * @param  int     $search_level
+     * @return array{0: PhpParser\Node\Expr\Assign|PhpParser\Node\Expr\AssignOp|
+     PhpParser\Node\Expr\AssignRef|null, 1: int}
+     */
+    private function findAssignExp(
+        PhpParser\Node\Expr $current_node,
+        string $var_id,
+        int $var_start_loc,
+        int $search_level = 1
+    ) {
+        if ($current_node instanceof PhpParser\Node\Expr\Assign ||
+            $current_node instanceof PhpPArser\Node\Expr\AssignOp ||
+            $current_node instanceof PhpParser\Node\Expr\AssignRef) {
+            $var = $current_node->var;
+
+            if ($var instanceof PhpParser\Node\Expr\Variable &&
+                $var->name === substr($var_id, 1) &&
+                $var->getStartFilePos() === $var_start_loc) {
+                return [$current_node, $search_level];
+            }
+
+            $rhs_exp = $current_node->expr;
+            $rhs_search_result = $this->findAssignExp($rhs_exp, $var_id, $var_start_loc, $search_level + 1);
+            return [$rhs_search_result[0], $rhs_search_result[1]];
+        } else {
+            return [null, $search_level];
+        }
+    }
+
+    /**
+     * @param  string   $var_id
+     * @param  CodeLocation $var_loc
+     * @return bool
+     */
+    private function checkIfVarRemoved($var_id, CodeLocation $var_loc): bool
```

Comment:
this can have a regular `string $var_id` typehint (and your can then remove the docblock, as it's extraneous).